src/main/java/datawave/microservice/query/storage/CachedQueryStatus.java

```diff
@@ -0,0 +1,445 @@
+package datawave.microservice.query.storage;
+
+import java.util.Set;
+import java.util.Timer;
+import java.util.TimerTask;
+
+import org.apache.accumulo.core.security.Authorizations;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+import datawave.core.query.configuration.GenericQueryConfiguration;
+import datawave.core.query.logic.QueryKey;
+import datawave.microservice.authorization.user.DatawaveUserDetails;
+import datawave.microservice.query.Query;
+import datawave.microservice.querymetric.BaseQueryMetric;
+import datawave.webservice.query.exception.DatawaveErrorCode;
+
+/**
+ * This class will cache a QueryStatus object for a specified query. The underlying storage will be polled at a specified interval, or on demand if the current
+ * object is deemed too old. If it is desired that the storage is polled at a specified interval, then use startTimer after construction. Otherwise the query
+ * status will be updated when a getter is called iff the current instance is too old. All setter methods except for numResultsReturned, numResultsGenerated,
+ * nextCount, and seekCount will force an update and will update the underlying storage. Updates made for the counters specified will only be updated in the
+ * underlying storage when the cache is refreshed per the specified interval.
+ */
+public class CachedQueryStatus extends QueryStatus {
+    private final QueryStorageCache cache;
+    private final String queryId;
+    private final long invalidCacheMs;
+    
+    private QueryStatus queryStatus = null;
+    private volatile long queryStatusTimeStamp = -1;
+    private Timer timer = null;
+    
+    public CachedQueryStatus(QueryStorageCache cache, String queryId, long invalidCacheMs) {
+        this.cache = cache;
+        this.queryId = queryId;
+        this.invalidCacheMs = invalidCacheMs;
+        loadQueryStatus();
+    }
+    
+    /**
+     * Update the query status, merging in any pending updates and executing the supplied query status updater
+     * 
+     * @param updater
+     */
+    public synchronized void updateQueryStatus(Runnable updater) {
+        QueryStorageLock lock = cache.getQueryStatusLock(queryId);
+        lock.lock();
+        try {
+            forceCacheUpdateInsideSetter();
+            updater.run();
+            cache.updateQueryStatus(queryStatus);
+        } finally {
+            lock.unlock();
+        }
+    }
```

Comment:
Is there a use case for a bulk update method? I could see a scenario where multiple updates are made one after the other, and that would cause repeated locking and unlocking. 

If the situation I'm describing wouldn't arise, please ignore me.

Reply:
What would a bulk update look like to you?  The way I understand it, the usage would be to pull the query status from the cache, make all of the updates you need to make to that local copy, and then as a locked operation push that back out to the cache to update the status.  For certain variables stored in the status, they might be updated by multiple services at the same time, and in that case we should lock the query id, pull the status, update the status (quickly), and then push it back out and unlock to ensure that the operation is atomic.  

I guess I need to understand more about the scenario you're thinking about.

Reply:
Based on my read of the setter methods that utilize the updateQueryStatus method, I could see a situation where the client is trying to update multiple setters at once within the context of processing a result page.
```java
// new page returned
cachedStatus.setNumResultsReturned()
cachedStatus.setLastPageNumber()
cachedStatus.setLastUsedMillis()
```
Each of those calls will trigger a lock/update/unlock cycle. It would make more sense to me if the flow was something like this
```java
// new page returned

// set the increment values without touching the cache
cachedStatus.setNumResultsReturned()
cachedStatus.setLastPageNumber()
cachedStatus.setLastUsedMillis()

cachedStatus.forceCacheUpdate()  // single lock, multiple variables updated
```

Reply:
Ahh ok, I see.  yeah, in my opinion it should be
 - lock the cache for that id
 - get the latest status from the cache
 - update the values for those three things in the cache
 - push the status back into the cache
 - unlock the cache

i created a helper class which lets you pass a lambda to update the values in the status in a locked fashion.  i would prefer to see that used instead of this class.  

The class/method is here: 
https://github.com/NationalSecurityAgency/datawave-spring-boot-starter-query/blob/feature/queryMicroservices/src/main/java/datawave/microservice/query/util/QueryStatusUpdateUtil.java#L60